Prevent saving the enterprise fee when a per order calculator is selected along with 'Inherit from product'

abdellani opened the pull request:
#### What? Why?

- Closes #10348

<!-- Explain why this change is needed and the solution you propose.
     Provide context for others to understand it. -->



#### What should we test?
<!-- List which features should be tested and how.
     This can be similar to the Steps to Reproduce in the issue.
     Also think of other parts of the app which could be affected
     by your change. -->

- Go to the Enterprise Fees page
- Selected an existing Enterprise Fee ( or you  can create a new one )
- select 'Inherit From Product' as the tax category
- select any 'Per Order' calculator.
- Fill in the remaining field if you're creating a new enterprise fee.

You should see an error message:
""You cannot select 'Inherit From Product' when a per order calculator is selected""
#### Release notes

<!-- Please select one for your PR and delete the other. -->

Changelog Category: User facing changes

<!-- Choose a pull request title above which explains your change to a
     a user of the Open Food Network app. -->

The title of the pull request will be included in the release notes.


#### Dependencies
<!-- Does this PR depend on another one?
     Add the link or remove this section. -->



#### Documentation updates
<!-- Are there any wiki pages that need updating after merging this PR?
     List them here or remove this section. -->

jibees reviewed (approved):
Thanks for this one. I think it is worth creating a controller spec for this particular case, what do you think?

mkllnk reviewed (approved):
Good work. I added another commit to show how you can simplify the translation of the error message.

Before, the error message included the attribute name which looked odd:
![Screenshot from 2023-03-06 14-22-48](https://user-images.githubusercontent.com/3524483/223015003-ea8d6d14-a20e-4a7e-ac3d-8e61154a7add.png)

But now it's just the new message which I tested in the spec and manually with the dev server.

Another improvement would be to move this code to a validation method instead of a `before_save` action. It would then show this error message along with other validation errors. The existing code would need to be called before validation then as well. Feel free to open a pull request for that if you like to play around with it.

I'm moving this to test ready for now so that we get the bug fix in the next release.

abdellani commented:
@mkllnk 
This is new to me.
Thank you :)

filipefurtad0 commented:
Hey @abdellani,

I think we need a second thought on this one.

If we have a previously existing enterprise fee - even if set as per item:
- updating the page, i.e., creating or editing always triggers the error

![image](https://user-images.githubusercontent.com/49817236/223513701-f3cd5570-c987-4279-a1a2-69b9c38c59df.png)


Also, I've tested from a situation where we have no **previous enterprise fees**: the error seems to always get triggered, also when creating a fee with a calculator operating over items:

![image](https://user-images.githubusercontent.com/49817236/223514045-9a66f379-1a12-4d33-8baf-bfe85e6634e7.png)

In fact, clicking update with an empty form triggers the warning as well.

For this reason, I think we should move this one back to In dev for now.

-------------------------

In any case, for the next iteration, I think we need to consider the case of pre-existing fees applied at the order level + the "Inherit from product": 

![image](https://user-images.githubusercontent.com/49817236/223514887-09301152-055d-4cd1-aac2-7e4d3e7ad23a.png)

In this situation, the user will not be able to update this page, unless, previously created fees are changed; I'm not sure how to deal with these. I'll add some more comments to the issue. Moving back to In Dev for now.

filipefurtad0 commented on spec/models/enterprise_fee_spec.rb:
```diff
@@ -9,6 +9,19 @@
+      expect(subject.errors.full_messages.first).to eq(
+        "Inheriting the tax categeory requires a per-item calculator."
+      )
+    end
```
In addition to this one, I'd suggest to add an it block, in which the validation is not triggered and saving into the DB is successful.

dacook replied:
`±` Good suggestion 👍

dacook replied:
I think we should still add a test for the successful cases. I think the most important one is:

* Per-item calculator, with inheritance ticked

abdellani replied:
I updated the tests to check the per-order and per-item calculators.
Now, I'm thinking we only discussed the two types of calculators, how about the others?
![image](https://user-images.githubusercontent.com/16113997/230706415-686e0790-98db-499a-af23-b9b0c2109d3e.png)

dacook replied:
I can see in one of @filipefurtad0's [comments](https://github.com/openfoodfoundation/openfoodnetwork/issues/10348#issuecomment-1410664238) that these three can inherit tax category:

> Flat percent (per item), Flat rate (per item), Weight

So I assume all others cannot. Is that right Filipe?

abdellani replied:
if that's true, maybe we can extends the original issue to include all the categories (except per-item) ?

filipefurtad0 replied:
We've clarified it over a huddle with @abdellani - I think we're good: weight is indeed applied over a line item, but should be included in the PR already :+1

dacook replied:
Ok, sounds like we're all good!

filipefurtad0 reviewed (commented):
Also, maybe this one would be suitable to add a request spec in which:
- submitting a order fee + inherit option -> error message + HTTP status code 304 (Not Modified)
- submitting a item fee + inherit option -> no error message + HTTP status code 200

Would this make sense?

abdellani commented:
Hi @filipefurtad0 

I tested the two first scenarios and I was not able to reproduce the issue locally.

Screenshots:

>If we have a previously existing enterprise fee - even if set as per item:
>
>* updating the page, i.e., creating or editing always triggers the error


![image](https://user-images.githubusercontent.com/16113997/223644122-49c02dfa-bc8a-4c47-8953-b4800baf0705.png)


>Also, I've tested from a situation where we have no previous enterprise fees: the error seems to always get triggered, also when creating a fee with a calculator operating over items:


![image](https://user-images.githubusercontent.com/16113997/223644803-4a501c74-569f-452a-8313-9e2ad939fc7e.png)

Did I miss anything?

--------------

>In any case, for the next iteration, I think we need to consider the case of pre-existing fees applied at the order level + the "Inherit from product":


Since it's not a valid configuration, maybe we can reconfigure the existing enterprise fees to the closest valid configuration? or maybe remove them?

filipefurtad0 commented:
Thanks for checking @abdellani .

> I tested the two first scenarios and I was not able to reproduce the issue locally.

I'll test again, at staging environment, and locally as well, on a test environment.

filipefurtad0 commented:
Hey @abdellani,
 
In a test environment, indeed this is possible: I've created some test cases [here](https://github.com/openfoodfoundation/openfoodnetwork/pull/10539).

In staging however, I can't seem to create any fees, also in new enterprises. I've logged into the server and checked the logs, I think this might be happening because the validation occurs also over other previously created fees, for other enterprises:

![image](https://user-images.githubusercontent.com/49817236/224078252-3ed3cb80-15f3-4556-aaa6-144aa98a09d0.png)

Any ideas?

abdellani commented:
Hi @filipefurtad0 
I check the controller's code.

https://github.com/openfoodfoundation/openfoodnetwork/blob/62d0ce290a53ba616e9894a14c30e32de0f616a7/app/controllers/admin/enterprise_fees_controller.rb#L38-L47

It looks like `Sets::EnterpriseFeeSet` loads all the existing enterprise fees.

https://github.com/openfoodfoundation/openfoodnetwork/blob/62d0ce290a53ba616e9894a14c30e32de0f616a7/app/services/sets/enterprise_fee_set.rb#L3-L11

One solution is to update the existing enterprise fees that have an invalid configuration. What do you think?

filipefurtad0 commented:
Hey @abdellani 

> It looks like Sets::EnterpriseFeeSet loads all the existing enterprise fees.

:muscle: Thanks for verifying this!

> One solution is to update the existing enterprise fees that have an invalid configuration. What do you think?

Yes, indeed. We were discussing this yesterday with @RachL and @jibees, some thoughts:
- for the invalid cases we'd ideally replace the selected option "Inherit from product" with a tax category which needs to be associated to a zero tax rate
- in these cases, the `enterprise_fee` has `inherits_tax_category` = `true` -> it should be set to false
- this should concern only calculators with `calculable_type` = `"EnterpriseFee"` and with `type` like ` flat rate per order`, `flexible rate per order`, `price sack`

filipefurtad0 commented on db/migrate/20230316034707_disable_inherits_tax_category_from_per_order_enterprise_fees.rb:
```diff
@@ -0,0 +1,16 @@
+  def change
+    EnterpriseFee.where(id: enterpise_fee_ids)
+      .where(inherits_tax_category: true)
+      .update_all(inherits_tax_category: false)
```
We set these into `false`, but it is not clear to me (from the code) which tax category will be selected for the fee, in this case. Ideally, it would be something like "None" or category associated to a zero tax rate. Do you know what happens with this migration @abdellani ?

Otherwise, let's move to test ready and test it there. But I feel this is an acceptance criteria which we should not overlook on testing: we need to assure that the migration sets:

- inherits_tax_category: true
- the selected tax category has a zero tax rate associated

The risk here is, that a random or first option is selected, which can lead to incorrect tax rate attribution to enterprise fees. 

I can't tell much more from the code, s

filipefurtad0 commented:
Hey @abdellani ,

Deployment failed, it seems there is something with the migration. I'm transcribing the error below:

` StandardError · An error has occurred, this and all later migrations canceled: Rails attempted to auto-load: /home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-03-21-151843/app/models/enterprise_fee.rb Which is in your project's `app/` directory. The good_migrations gem was designed to prevent this, because migrations are intended to be immutable and safe-to-run for the life of your project, but code in `app/` is liable to change at any time. The most common reason for this error is that you may be referencing an ActiveRecord model inside the migration in order to use the ActiveRecord API to implement a data migration by querying and updating objects. For instance, if you want to access a model "User" in your migration, it's safer to redefine the class inside the migration instead, like this: class MakeUsersOlder < ActiveRecord::Migration class User < ActiveRecord::Base # Define whatever you need on the User beyond what AR adds automatically end def up User.find_each do |user| user.update!(:age => user.age + 1) end end def down #... end end For more information, visit: https://github.com/testdouble/good-migrations`

More details on [bugsnag](https://app.bugsnag.com/yaycode/openfoodnetwork-uk/errors/6419cb389b0ab60008bbe7c5?event_id=6419cb3700b9d3585fdc0000&i=sk&m=nw) and [semaphore](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/servers/uk-staging/deploys/1607).

Moving back to in dev, for now.

abdellani commented:
@filipefurtad0 

I guess the problem is that I referred to a constant, declared on the ActiveRecord Model, from the migration.
If someone edits the constant later, that migration will have a different behavior.
I'll replace that constant with its value. This should fix the problem.

filipefurtad0 commented:
I've tried staging (on a different server too) this new version @abdellani , but still get the [same error](https://semaphoreci.com/openfoodfoundation/openfoodnetwork-2/servers/fr-staging/deploys/973).

dacook commented:
@abdellani, in the deployment I see the message below.
I guess we never run migrations in CI (the test environment loads directly from the latest db/schema.rb version).
But I would have expected this to occur when running the the migration in your development environment, did you execute it there?

```
Rails attempted to auto-load:

/home/openfoodnetwork/apps/openfoodnetwork/releases-old/2023-03-22-114538/app/models/enterprise_fee.rb

Which is in your project's `app/` directory. The good_migrations
gem was designed to prevent this, because migrations are intended
to be immutable and safe-to-run for the life of your project, but
code in `app/` is liable to change at any time.

The most common reason for this error is that you may be referencing an
ActiveRecord model inside the migration in order to use the ActiveRecord API
to implement a data migration by querying and updating objects.

For instance, if you want to access a model \"User\" in your migration, it's safer
to redefine the class inside the migration instead, like this:

class MakeUsersOlder < ActiveRecord::Migration
  class User < ActiveRecord::Base
    # Define whatever you need on the User beyond what AR adds automatically
  end

  def up
    User.find_each do |user|
      user.update!(:age => user.age + 1)
    end
  end

  def down
    #...
  end
end

For more information, visit:

https://github.com/testdouble/good-migrations
```

abdellani commented:
Sorry @filipefurtad0 , @dacook 

I just tested the queries on `rails console`.
Now, I fixed the migration and tried `rails db:migrate`.

filipefurtad0 commented:
Ok @abdellani ! Just as an early feedback: The PR stages now :tada: 

It migrates the selected option in tax category: We go from this scenario (per order calculators with the `Inherit from product` option selected):

![image](https://user-images.githubusercontent.com/49817236/223514887-09301152-055d-4cd1-aac2-7e4d3e7ad23a.png)

To this scenario:

No tax category selected:

![image](https://user-images.githubusercontent.com/49817236/227179358-13ccfd51-d58d-40b4-bdaa-23861bdc8c0f.png)

So the migration works correctly :raised_hands: 

The selection is not mandatory in this screen (this is the current behavior) so I'd say we're good here :ok_hand: 

Also, I've tested creating a valid, additional fee - this is now possible, no error is triggered:

![image](https://user-images.githubusercontent.com/49817236/227180874-294b94f0-decb-4162-a32f-f6fef981d95f.png)

So, this all looks good. I'd still wait for code review before any additional tests or merging.

dacook commented on db/schema.rb:
```diff
@@ -291,6 +291,17 @@
     t.index ["enterprise_id", "variant_id"], name: "index_inventory_items_on_enterprise_id_and_variant_id", unique: true
   end
 
+  create_table "invoices", force: :cascade do |t|
+    t.bigint "order_id"
+    t.string "status"
+    t.integer "number"
+    t.jsonb "data"
+    t.date "date", default: -> { "CURRENT_TIMESTAMP" }
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.index ["order_id"], name: "index_invoices_on_order_id"
+  end
+
```
`±` Oops!
This happened because rails migrate always regenerates the schema.db file from your dev database.

dacook commented on db/migrate/20230316034707_disable_inherits_tax_category_from_per_order_enterprise_fees.rb:
```diff
@@ -0,0 +1,28 @@
+      .update_all(inherits_tax_category: false)
+  end
+
+  def enterpise_fee_ids
+    Spree::Calculator.where(
+      calculable_type: 'EnterpriseFee',
+      type: per_order_calculators
+    ).pluck(:calculable_id)
+  end
```
Not that efficiency matters here, I thought it worth pointing out that we could probably use a `join` to avoid the extra query.

mkllnk replied:
Yes, or at least a `select` to have a subquery.

abdellani replied:
I replaced the old code with a join

dacook reviewed (changes requested):
Almost ready, just a couple of updates to make.

mkllnk reviewed (commented):
There's a conflict on the db/schema.rb but it needs to be rewritten anyway. I agree with David's comments but otherwise all good.

drummer83 commented:
@filipefurtad0 Are you willing to continue/finish testing this one? I think it would take me quite some time to dive into this and get a test setup ready.

filipefurtad0 commented:
Hey @abdellani ,

I've deleted the previously ran migration from the migrations table, on the server - so I could re-run it:

```
openfoodnetwork@staging:~/apps/openfoodnetwork/current$ bundle exec rails db:migrate:status | grep 'inherit'
   up     20230316034707  Disable inherits tax category from per order enterprise fees
openfoodnetwork@staging:~/apps/openfoodnetwork/current$ psql -h localhost openfoodnetwork ofn_user
psql (10.17 (Ubuntu 10.17-1.pgdg16.04+1))
SSL connection (protocol: TLSv1.2, cipher: ECDHE-RSA-AES256-GCM-SHA384, bits: 256, compression: off)
Type "help" for help.

openfoodnetwork=> delete from schema_migrations where version = '20230316034707';
DELETE 1
openfoodnetwork=> \q
openfoodnetwork@staging:~/apps/openfoodnetwork/current$ bundle exec rails db:migrate:status | grep 'inherit'
  down    20230316034707  Disable inherits tax category from per order enterprise fees

```

Then staged the PR, and got:
```

openfoodnetwork@staging:~/apps/openfoodnetwork/current$ bundle exec rails db:migrate:status | grep 'inherit'
   up     20230316034707  Disable inherits tax category from per order enterprise fees

```

But somehow, I did not observe any changes, as previously reported:

I still see the `inherit from product` option selected, on enterprise fees:

![image](https://user-images.githubusercontent.com/49817236/233073714-fc65b29b-be66-40fe-806b-c0e634afb447.png)

Any idea why?

abdellani commented:
@filipefurtad0 

I found the issue.

In the migration, we have two class declarations: `EnterpriseFee` and  `Spree::Calculator`.

As the association is polymorphic, we use `calculable_id` and `calculable_type` on the calculator to find the associated EnterpriseFee.

When I tried  `EnterpriseFee.joins(:calculator)`, the SQL query ran with `calculable_type= DisableInheritsTaxCategoryFromPerOrderEnterpriseFees::EnterpriseFee ` instead of `EnterpriseFee`.

I solved the issues by adding a new condition on the `Spree::Calculator` query and avoiding declaring the relationship as if it's not polymorphic.

sigmundpetersen commented:
Pushing this one through a short re-review

dacook commented on db/migrate/20230316034707_disable_inherits_tax_category_from_per_order_enterprise_fees.rb:
```diff
@@ -0,0 +1,29 @@
+
+class DisableInheritsTaxCategoryFromPerOrderEnterpriseFees < ActiveRecord::Migration[6.1]
+  class EnterpriseFee < ApplicationRecord
+    has_many :calculator, class_name: "Spree::Calculator", foreign_key: :calculable_id
```
From what I can understand in the last comment, you've updated this line. 
I think that using `as: :calculable` as [in the model](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/models/concerns/calculated_adjustments.rb#L9) would have achieved the same result (rails would guess "calculable_id" from "calculable").
But it sounds like this works too 👍 .

I think sometimes it's actually simpler and safer to just build the SQL directly for a migration, rather than try to re-create the model.

abdellani replied:
>I think that using as: :calculable as [in the model](https://github.com/openfoodfoundation/openfoodnetwork/blob/master/app/models/concerns/calculated_adjustments.rb#L9) would have achieved the same result (rails would guess "calculable_id" from "calculable").

@dacook that's correct, but the problem is with calculable_type. The SQL query will have the two following parameters 
1. calculable_id: will have the correct id of the enterprise fee
2. calculable_type: will be set to `DisableInheritsTaxCategoryFromPerOrderEnterpriseFees::EnterpriseFee`, which is wrong ( it must be `EnterpriseFee`)

dacook commented on db/migrate/20230316034707_disable_inherits_tax_category_from_per_order_enterprise_fees.rb:
```diff
@@ -0,0 +1,29 @@
+    end
+  end
+
+  def change
```
Migrations have two actions: `rake db:migrate` (perform the migration) and `rake db:rollback (undo the migration if possible). 

This method migrates the data to the new state, so it should only run on `migrate`. 
In this case we're not losing any information, so we don't need a rollback (I think the `inherits_tax_category` did nothing before).

The `change` method runs for both `migrate` and `rollback`. To only run on `migrate`, we can just define `up`:

```diff
-  def change
+  def up
```

This is only for future reference. It doesn't really matter in this case if it runs on rollbacks so I don't suggest we bother updating the code.

dacook reviewed (approved):
The update looks ok to me.
I noticed another issue with the migration code, but I believe it's not essential and we should just move forward.

Let's re-test.

In order to re-try the migration, we'll need to rollback or delete the migrations log like Filipe did earlier.

filipefurtad0 commented:
Hey @abdellani ,

This looks good now. Setting up fees with per-order calculators, with the `Inherit from product` tax category, **before this PR**:

![image](https://user-images.githubusercontent.com/49817236/235379180-a2539847-7c44-44ae-a566-74e3f401572d.png)

**After this PR**:
![image](https://user-images.githubusercontent.com/49817236/235379174-1420d055-ac66-4093-a7b8-f4934f9fcbba.png)

1) We can see that these have been successfully migrated, to have no selected tax category  :heavy_check_mark: 

2) Creating a new fee or editing existing ones is possible  :heavy_check_mark: 

3) Creating a new fee with per-order calculator and the `Inherit from product` tax category, triggers the error:
![image](https://user-images.githubusercontent.com/49817236/235379237-3b782bc7-455f-423c-b4ee-eac69ce61bf6.png)

The fee is not created :heavy_check_mark: 

4) By editing a fee, it is still possible to save the undesired combination into the DB: the error is triggered, but the DB is updated. 

For example, by changing:
![image](https://user-images.githubusercontent.com/49817236/235379517-5ba2663e-12ac-4fb1-98ca-9661bc69f340.png)
to
![image](https://user-images.githubusercontent.com/49817236/235379529-a6e4980c-7488-472e-8263-f4df01f59acb.png)

@abdellani , I did not think of this scenario before... the bug report only mentions fee creation. It would be great to merge this, to prevent further iterations. So in this case, I'd open an issue after merging.
On the other hand, I think that undesired combinations are possible, so we don't need to re-run a migration in production again, in this regard.

What do you think it is best?

abdellani commented:
@filipefurtad0 

I tried to reproduce the issue but I failed.
Can you please check the screencast? did I miss something?
https://user-images.githubusercontent.com/16113997/235596633-733c3e83-0e19-44f4-8c61-fc61bea099a6.mp4

filipefurtad0 commented:
Hey @abdellani ,

The steps I took to reproduce were:

1) Creating a fee with an allowed combination:
Tax Category: `Alberta GST`
Calculator: `Flat percent (per Item)`

![image](https://user-images.githubusercontent.com/49817236/235627305-fac45d27-eb67-430e-923c-6218dc3125fa.png)

2) Refresh the page to verify it is saved into the DB

3) Change the Tax category to be inherited from product - notice this combination is allowed, but the UI displays the warning:
Tax Category: `Inherit from product`
Calculator: `Flat percent (per Item)`

![image](https://user-images.githubusercontent.com/49817236/235628799-f274333f-b8d3-4c54-9fd1-71695d9dfd83.png)

4) Change the calculator to an order based calculator - notice the warning here as well -> this is correct
Tax Category: `Inherit from product`
Calculator: `Flat percent (per Order)`

![image](https://user-images.githubusercontent.com/49817236/235629384-53fdcaab-5de2-4e2e-9fb0-f16077c8faab.png)

5) Refresh the page or check the DB to verify the not allowed combination was stored


I see two issues here:
- on point 3) ideally, we would not display the warning
- on point 4) the not allowed combination is stored on the DB, despite the warning

There is a lot going on on this PR. Do you think we should split this in two? For example:
- PR 1 -> addressing all UI issues
- PR 2 -> after merging PR 1, perform the DB migration

What do you think?

abdellani commented:
Hi @filipefurtad0 

The order of the steps on my test is different that's why I didn't detect the problem.

Suppose we start from any valid state, for example:
1. inheritance from product: disabled
2. calculator: flat percent per order


On my test:
1. I changed enabled the inheritance and changed the calculator at the same time.

Results:
1. the calculator will be replaced
2. the inheritance flag will not change because the combination is invalid


On your test:
1. you enabled the inheritance first
2. you changed the calculator

results:
1. the calculator will be replaced
2. the inheritance flag will be set


The reason behind this behavior:

After the update button is clicked, the backend:
1. replace the calculator (remove the old calculator and create the selected one) 
2. update the enterprise fee
3. run the validations ( which's too late because the calculator is already replaced)

I'm not against splitting the PR, but I think that the problem is with the validations.

Now, I'm not sure about the best way to solve the issue.

I tried to prevent the creation of the selected calculator under invalid conditions but didn't solve the problem because the original calculator will be removed first.

The bulk update is done through `ModelSet`, I don't think that it supports adding validations before updating the enterprise fees collection.


I'm thinking about introducing parameter validation on `enterprise_fee_controller`, but is there a better solution?

dacook commented:
It sounds like the problem is that the association is updated regardless of the validation. I think this is because associations in Rails are saved immediately. 
Hm...

It would be nice to solve it in the model, but the only way I can think to do that is move the validation to the "per order" calculator models, relying on the assoicated fee 'inherits' value. 

So perhaps we need to pre-validate in the controller or ModelSet. I'm not familiar with  ModelSet and I'm guessing it's not flexible enough to do that. Perhaps you could add a hook for that, otherwise go ahead and validate in the controller.

abdellani commented:
> I agree. For example, on point 3. https://github.com/openfoodfoundation/openfoodnetwork/pull/10512#issuecomment-1531161376 the banner is displayed, although the combination is legitimate.

@filipefurtad0  I'm not able to reproduce it.
Can you please check the screenshot?

https://user-images.githubusercontent.com/16113997/237044125-a1329a3b-339d-4a88-9729-f71ba611a252.mp4


> So perhaps we need to pre-validate in the controller or ModelSet. I'm not familiar with ModelSet and I'm guessing it's not flexible enough to do that. Perhaps you could add a hook for that, otherwise go ahead and validate in the controller.

@dacook Ok,  I'll go with the second option.

abdellani commented:
@dacook 

We already have a validation on the controller, it was introduced a few months ago https://github.com/abdellani/openfoodnetwork/commit/755107ec9f60034119cbabdf8c2b077aedd92fee

abdellani commented:
@filipefurtad0 
I run the test again.
I think it's good now.

mkllnk commented on app/controllers/admin/enterprise_fees_controller.rb:
```diff
@@ -2,11 +2,13 @@
 
 require 'open_food_network/order_cycle_permissions'
 
+# rubocop:disable Metrics/ClassLength
```
We should add the filename to the rubocop todo file instead of disabling the cop here.

dacook reviewed (approved):
Good one!👍

filipefurtad0 commented:
Hey @abdellani ,

I think we're good now!! :tada: 

Before this PR:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/77c2f3d0-aff2-4fd7-8ff3-cf9848607ed3)

After this PR:

![image](https://github.com/openfoodfoundation/openfoodnetwork/assets/49817236/b097cc2b-25cb-4d29-b3c4-843cd14ecb85)

We can see, that for the order level calculators, the Inherit from product option was removed :heavy_check_mark: 

Also, I was unable to create new fees with invalid combinations, either by creating or editing existing fees (the test cases previously mentioned).

This is ready to go!

Congratulations for the work and resilience :clap: :clap: :clap: 

PS: I'll update the tests on #10539.
